### PR TITLE
chore(windows): removed cached context windows engine

### LINKED ISSUE #10052: chore(windows): remove cached context management from Keyman Engine
Part of #9999.

- [x] Remove `AITIP::MergeContextWithCache`
- [x] Remove `AITIP::CopyContext`, `AITIP::RestoreContextOnly`
- [x] Remove `AITIP::QueueAction` (see #10050)
- [x] Remove `AIWin2000Unicode::context`
- [x] Where `context->Reset()` is used, instead tell core to reset `km_core_state_context_clear`
- [x] Move `AppContext` class to calldll.cpp, and remove from appint.h, and remove unused functionality (as AppContext still used by some of the calldll interactions)
- [x] Remove `AIWin2000Unicode::QueueAction`
- [x] Remove `ContextItemsFromAppContext` from appint
- [x] Update `AppIntegration::ReadContext()` to `bool AppIntegration::ReadContext(WCHAR *buf)`
- [x] In kmprocess.cpp, ReadContext update: if it returns TRUE, then we will `km_core_state_context_set_if_needed`

### PR DESCRIPTION
**Pull request by rc-swag**
Fixes #10052 

# User Testing

I have taken the context and store test for the acceptance tests on the wiki. We don't need the stores will just use the 
 **043 - output and keystroke.kmx** and **045 - deadkey and context.kmx**
Use the text editor available on this PR in the User Testing section

- GROUP_TEXT_EDITOR
- GROUP_WORD_PAD  (TSF)

#### Test cases

<details>
<summary>click to expand</summary>

The key referred to as "enter" is alternatively labelled "return" on some keyboards.

Keyboards for the following test are found in the zip file [store_context_kbd.zip](https://github.com/keymanapp/keyman/files/9441852/store_context_kbd.zip)

**TEST_OUTPUT_KEYSTROKE**
1. Install **043 - output and keystroke.kmx**
2. Open `Text Editor`
3. Select  043 - output and keystroke.kmx as the current keyboard.
4. Type <kbd>1</kbd><kbd>2</kbd><kbd>3</kbd>

Expected Result: "abd3"

**TEST_OUTPUT_KEYSTROKE_INVALID_CONTEXT**

1. If not installed from the previous test - Install **043 - output and keystroke.kmx**
2. Open `Text Editor`
3. Select  043 - output and keystroke.kmx as the current keyboard.
4. Type <kbd>1</kbd><kbd>2</kbd><kbd>Enter</kbd><kbd>3</kbd>

Expected Result: \
"abc\
3"

**TEST_OUTPUT_KEYSTROKE_CONTEXT_FRAME_KEY**

1. If not installed from the previous test - Install **043 - output and keystroke.kmx**
2. Open `Text Editor`
3. Select  043 - output and keystroke.kmx as the current keyboard.
4. Type <kbd>1</kbd><kbd>2</kbd><kbd>left arrow key</kbd><kbd>right arrow key</kbd><kbd>3</kbd>

Expected Result WordPad: "abd3"

Expected Result TextEditor: "abc3"

**TEST_DEADKEY_AND_CONTEXT**
1. Install **045 - deadkey and context.kmx**
2. Open `Text Editor`
3. Select  045 - deadkey and context.kmx as the current keyboard.
4. Type <kbd>y</kbd><kbd>z</kbd><kbd>Shift</kbd>+<kbd>/</kbd>

Expected Result: "correct"

**TEST_DEAD_CONTEXT_FRAME_KEY**
1. Install **045 - deadkey and context.kmx**
2. Open `Text Editor`
3. Select  045 - deadkey and context.kmx as the current keyboard.
4. Type <kbd>y</kbd><kbd>z</kbd> <kbd>left arrow key</kbd><kbd>right arrow key</kbd><kbd>Shift</kbd>+<kbd>/</kbd>

Expected Result: "\<ab\>?"
</details>

### CODE REVIEW

File: windows/src/engine/keyman32/appint/aiTIP.h [hunk 1]
```diff
@@ -50,28 +47,18 @@ class AITIP : public AIWin2000Unicode
...
 
 	virtual BOOL CanHandleWindow(HWND ahwnd);
 	virtual BOOL IsUnicode();
 
 	/* Context functions */
 
-	virtual void ReadContext();
+ /**
+   * Reads the current application context upto MAXCONTEXT length into the supplied buffer.
+   * @param  buf      The data buffer to copy current application context
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/appint/aiTIP.h:59]:
```suggestion
   * @param  buf      The data buffer to copy current application context into, must
   *                  be MAXCONTEXT WCHARs or larger.
```

File: windows/src/engine/keyman32/appint/aiWin2000Unicode.cpp [hunk 2]
```diff
@@ -87,33 +80,23 @@ BOOL AIWin2000Unicode::IsUnicode()
...
-{
-	return context->Buf(n);
-}
-
-WCHAR *AIWin2000Unicode::ContextBufMax(int n)
-{
-	return context->BufMax(n);
+BOOL AIWin2000Unicode::ReadContext(PWSTR buf) {
+  UNREFERENCED_PARAMETER(buf);
+  return FALSE;
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/appint/aiWin2000Unicode.cpp:85]:
```suggestion
  // We cannot read any context from legacy apps, so we return a 
  // failure here -- telling Core to maintain its own cached 
  // context.
  return FALSE;
```

File: windows/src/engine/keyman32/kmprocess.cpp [hunk 3]
```diff
@@ -70,23 +70,30 @@
...
+  if (!_td || !_td->lpActiveKeyboard || !_td->lpActiveKeyboard->lpCoreKeyboardState){
+        return "";
+  }
 
+  WCHAR buf[(MAXCONTEXT * 3) + 1];  // *3 if every context item was a deadkey
+  km_core_context_item *citems = nullptr;
+
+  if (KM_CORE_STATUS_OK != km_core_context_get(
+    km_core_state_context(_td->lpActiveKeyboard->lpCoreKeyboardState), &citems)) {
+    km_core_context_items_dispose(citems);
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/kmprocess.cpp:85]:
```suggestion
```

On failure here, citems will still be nullptr

File: windows/src/engine/keyman32/kmprocess.cpp [hunk 4]
```diff
@@ -98,14 +105,15 @@ char *getcontext_debug() {
...
-  if (KM_CORE_STATUS_OK != km_core_context_set(km_core_state_context(_td->lpActiveKeyboard->lpCoreKeyboardState), citems)) {
-    km_core_context_items_dispose(citems);
-    return FALSE;
+  WCHAR application_context[MAXCONTEXT];
+  if (_td->app->ReadContext(application_context)) {
+    km_core_context_status result;
+    result = km_core_state_context_set_if_needed(_td->lpActiveKeyboard->lpCoreKeyboardState, reinterpret_cast<const km_core_cp *>(&application_context));
+    if (result == KM_CORE_CONTEXT_STATUS_ERROR || result == KM_CORE_CONTEXT_STATUS_INVALID_ARGUMENT) {
+      SendDebugMessageFormat(0, sdmGlobal, 0, "ProcessEvent SetContext if needed Result:False %d ", FALSE);
+    }
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/kmprocess.cpp:114]:
```suggestion
    if (result == KM_CORE_CONTEXT_STATUS_ERROR || result == KM_CORE_CONTEXT_STATUS_INVALID_ARGUMENT) {
      SendDebugMessageFormat(0, sdmGlobal, 0, "Process_Event_Core: km_core_state_context_set_if_needed returned %d", result);
    }
```

**Review by mcdurdin — APPROVED** [review]
LGTM

**Review by mcdurdin — APPROVED** [review]

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_TEXT_EDITOR: 

* **TEST_OUTPUT_KEYSTROKE (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Unzipped then installed the store_context keyboards using Keyman Configuration dialog. 2. Installed Text Editor 64 (EC Test App) from the PR build then Opened it. 3. Selected 043 - output and keystroke.kmx as the current keyboard. 4. Typed "123". Verified that "abd3" appeared on the Text Editor. 


* **TEST_OUTPUT_KEYSTROKE_INVALID_CONTEXT (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Opened the Text Editor (64 bit). 2. Selected 043 - output and keystroke.kmx as the current keyboard. 3. Typed "12". Pressed te <Enter> key. Typed "3". Verified that 
"abc
3" appeared on the Text Editor.


* **TEST_DEADKEY_AND_CONTEXT (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Opened the Text Editor (64 bit). 2. Selected 045 - deadkey and context.kms as the current keyboard. 3. Typed "yz shift+/". Verified that the text "correct" appear as the output. 


* **TEST_DEAD_CONTEXT_FRAME_KEY (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Opened the Text Editor (64 bit). 2. Selected 045 - deadkey and context.kms as the current keyboard. 3. Typed "yz lef arrow key right arrow key shift+/". Verified that "<ab>?" text appear as the output. 

..Text Editor Output

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_WORD_PAD: (TSF)

* **TEST_OUTPUT_KEYSTROKE (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Unzipped then installed the store_context keyboards using Keyman Configuration dialog. 2. Opened WordPad. 3. Selected 043 - output and keystroke.kmx as the current keyboard. 4. Typed "123". Verified that "abd3" appeared on the Text Editor. 

* **TEST_OUTPUT_KEYSTROKE_INVALID_CONTEXT (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Opened WordPad. 2. Selected 043 - output and keystroke.kmx as the current keyboard. 3. Typed "12". Pressed te <Enter> key. Typed "3". Verified that 
"abc
3" appeared on the Text Editor.  


* **TEST_DEADKEY_AND_CONTEXT (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Opened WordPad. 2. Selected 045 - deadkey and context.kms as the current keyboard. 3. Typed "yz shift+/". Verified that the text "correct" appears as the output.

* **TEST_DEAD_CONTEXT_FRAME_KEY (PASSED):** Tested with the attached PR build (keyman 17.0.216-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Opened WordPad. 2. Selected 045 - deadkey and context.kms as the current keyboard. 3. Typed "yz lef arrow key right arrow key shift+/". Verified that "<ab>?" text appears as the output. 

..in WordPad Editor

**Comment by mcdurdin** [comment]
> FYI, I got an error message during install the attached PR build (Keyman 17.0.216-alpha-test-10065) on Windows 10 OS. When I press the "Retry" button multiple times, the installation process proceeds further. Just want to bring it to your attention.

This is the same issue as you saw earlier -- issues with the keyman.com servers being slow. I am investigating.

**Comment by rc-swag** [comment]
# Test Results

### GROUP_TEXT_EDITOR: 

* **TEST_OUTPUT_KEYSTROKE_CONTEXT_FRAME_KEY (PASS):** notes

### GROUP_WORD_PAD: (TSF)

* **TEST_OUTPUT_KEYSTROKE_CONTEXT_FRAME_KEY (FAILED):**  The result is `abd3` which suggests getting the context from TSF apps and setting it in the core is failing.

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_WORD_PAD: (TSF)

* **TEST_OUTPUT_KEYSTROKE_CONTEXT_FRAME_KEY (PASSED):** Tested with the attached PR build (Keyman 17.0.220-alpha-test-10065) on Windows 10 OS and here is my observation: 1. Installed the 043 - output and keystroke.kmx keyboard. 2. Opened the WordPad app. 3. Followed the steps mentioned in the instructions and show the expected result. 

..in WordPad

**Pull request merged**